### PR TITLE
commands: repo fsck

### PR DESCRIPTION
**Pull request by pfista**
This adds a repo fsck command as outlined in #2457.

### CODE REVIEW

File: core/commands/repo.go [hunk 1]
```diff
@@ -152,3 +158,49 @@ RepoPath        string the path to the repo being currently used
 		},
 	},
 }
+
+var repoFsckCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "Removes repo lockfiles",
+		ShortDescription: `
+'ipfs repo fsck' is a plumbing command that will remove repo and level db lockfiles
```
Comment (RichardLitt) [inline core/commands/repo.go]: Add `.` at the end.

File: core/commands/repo.go [hunk 2]
```diff
@@ -152,3 +158,49 @@ RepoPath        string the path to the repo being currently used
...
+			return
+		}
+
+		dsLockFile := filepath.Join(dsPath, "LOCK") // TODO: get this lockfile programmatically
+		repoLockFile := filepath.Join(configRoot, lockfile.LockFile)
+
+		log.Infof("Removing repo lockfile: %s", repoLockFile)
+		log.Infof("Removing datastore lockfile: %s", dsLockFile)
+
+		os.Remove(repoLockFile)
```
Comment (whyrusleeping) [inline core/commands/repo.go]: we will definitely want to check the error returned here.

File: core/commands/repo.go [hunk 3]
```diff
@@ -152,3 +158,49 @@ RepoPath        string the path to the repo being currently used
...
+		os.Remove(dsLockFile)
+
+		s := fmt.Sprintf("Lockfiles have been removed.")
+		log.Info(s)
```
Comment (whyrusleeping) [inline core/commands/repo.go]: can just do `log.Info("Lockfiles....`, drop the Sprintf step.

File: core/commands/repo.go [hunk 4]
```diff
@@ -152,3 +158,49 @@ RepoPath        string the path to the repo being currently used
...
+		ShortDescription: `
+'ipfs repo fsck' is a plumbing command that will remove repo and level db lockfiles
+`,
+	},
+	Run: func(req cmds.Request, res cmds.Response) {
+
+		configRoot := req.InvocContext().ConfigRoot
+
+		// acquire the repo lock before removing it
+		repo, err := fsrepo.Open(configRoot)
```
Comment (whyrusleeping) [inline core/commands/repo.go]: I don't know if this will work correctly... i think the reason we want to remove the lockfiles is because theyre in a bad state and we *cant* take the locks.

We should however try and check if the daemon is online.

File: core/commands/repo.go (hunk 3 above)
Comment (pfista) [inline core/commands/repo.go]: @whyrusleeping what about `res.SetOutput`? Don't we need to return a MessageOutput? Or can you make a subcommand's type nil?

File: core/commands/repo.go (hunk 4 above)
Comment (pfista) [inline core/commands/repo.go]: @whyrusleeping To check for daemon status would I just make sure `req.InvocContext().GetNode()` doesn't return an error?

File: core/commands/repo.go (hunk 3 above)
Comment (whyrusleeping) [inline core/commands/repo.go]: ah, then just remove the `fmt.Sprintf` wrapper. Its not necessary, `s := "Lockfiles...."` works

File: core/commands/repo.go (hunk 4 above)
Comment (whyrusleeping) [inline core/commands/repo.go]: take the node you get back from `req.InvocContext().GetNode()` and check `n.OnlineMode()` on it.

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 33d1037bba2d899c01152ca293b08ba8fb4b194d
  * Invalid signoff. Commit message must end with  
    **License: MIT  
Signed-off-by: .* <.*>**

We ask for a few features in the commit message for Open Source licensing hygiene and commit message clarity.
`git commit --amend` can often help you quickly improve the commit message.
[Guidelines](https://github.com/ipfs/community/blob/master/contribution-guidelines.md#commit-messages) and [a script](https://github.com/ipfs/community/blob/master/dev/tools/hooks/setup_commit_msg_hook.sh) are available to help in the long run.
Your feedback on GitCop is welcome on [this issue](https://github.com/ipfs/infrastructure/issues/23).

---
This message was auto-generated by https://gitcop.com

File: core/commands/repo.go [hunk 5]
```diff
@@ -152,3 +156,56 @@ RepoPath        string the path to the repo being currently used
...
+'ipfs repo fsck' is a plumbing command that will remove repo and level db lockfiles.
+`,
+	},
+	Run: func(req cmds.Request, res cmds.Response) {
+		n, err := req.InvocContext().GetNode()
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+		if !n.OnlineMode() {
```
Comment (whyrusleeping) [inline core/commands/repo.go]: so in this case, we want to make sure that we are *offline*. If we try running a fsk while the daemon is running, we would end up deleting the valid locks that the running daemon is holding

File: core/commands/repo.go [hunk 6]
```diff
@@ -152,3 +156,56 @@ RepoPath        string the path to the repo being currently used
...
+		}
+
+		dsLockFile := filepath.Join(dsPath, "LOCK") // TODO: get this lockfile programmatically
+		repoLockFile := filepath.Join(configRoot, lockfile.LockFile)
+
+		log.Infof("Removing repo lockfile: %s", repoLockFile)
+		log.Infof("Removing datastore lockfile: %s", dsLockFile)
+
+		err = os.Remove(repoLockFile)
+		if err != nil {
```
Comment (whyrusleeping) [inline core/commands/repo.go]: so, one here for both of these removes that we are okay with getting is the 'file does not exist' type errors. The `os` package gives us a nice way to check this:
```go
if err != nil && !os.IsNotExist(err) {
  res.SetError(err, cmds.ErrNormal)
  return
}
```

If the lockfiles don't exist, we're just fine with that, its not an error.

**Comment by whyrusleeping** [comment]
Okay, so i thought it was going to be difficult to do this, but then i found something cool.

If you rename the command from `repoFsckCmd` to `RepoFsckCmd` that 'exports' the command (makes it public). Then if you go to the very end of `cmd/ipfs/ipfs.go` and add
```go
	commands.RepoFsckCmd:       {cannotRunOnDaemon: true},
```

to the map there, it should have the desired effect of not running while the daemon is online.

**Comment by whyrusleeping** [comment]
make sure to test the various different scenarios (should probably add a sharness test)

- daemon is running
  - command should not run
- `~/.ipfs/api` exists, but the address inside points to a dead daemon.
  - command should run fine, and remove the api file
- daemon is not running, but `~/.ipfs/repo.lock` exists and has a non-zero size.
  - command should run and clean up that file.

**Comment by pfista** [comment]
Building out the fsck tests brought up some issues and I'd like some input
please.

## `CannotRunOnDaemon` reliance on API file

@whyrusleeping suggested using The `ipfs repo` command property
`cannotRunOnDaemon` which is a more elegant solution than checking if the daemon
is running inside the `fsck` subcommand. The problem is that this command
property currently only checks if the daemon is running iff the `api` file is
present.  This is a problem because if the api file isn't present and the daemon
is actually running, this command will still be executed when it shouldn't be.

I see two options moving forward and would like input on which to choose:
1. Just make the tests conform to the fact that an API file must be present for
   this property check.
2. Remove cannotRunOnDaemon's dependency on the existence of an API file to run.

Which direction should I take?

## non-zero repo.lock 

In what case will repo.lock be non-zero? Locks implementation states that if the
lockfile already exists, it must have a zero size or an error is returned. Lock
files should neither be read from nor written to. If it isn't an empty file,
`lock.go` will fail to acquire the lock leading to unexpected behavior for the
`fsck` command. This leads to the false-positive assumption that the daemon is
running because a lock couldn't be acquired when it should have been.

This all stems from using the command property `cannotRunOnDaemon`. 

At first, My tests used a repo.lock file that was a non-zero size and many of my
tests failed because of it. Because it was non-zero, `lock.go` returned the
aforementioned error so the daemon appeared to be running even though it wasn't.
Then, the `fsck` command wouldn't run because it's property `cannotRunOnDaemon`
is set to true. 

So I won't be able to sucessfully test for this repo.lock non-zero case unless:
1. `lock.go` doesn't care about non-zero lock files, and can still acquire the
   lock (Probably shouldn't do this)
2. I find a different method (not using the commmand property
   `cannotRunOnDaemon`) of checking if the daemon is running or not before
   running the command. 
3. OR, just ignore it. Don't test for this case because non-zero repo.lock files
   won't ever happen.

Below, scenario 1 cannot be tested for since it fails, while scenario 2 passes.

####  Scenario 1
Status
repo.lock exists and is non-zero
datastore/LOCK exists and is empty
api exists and is non-zero 
daemon is not running

Expected result:
all lock files are removed successfully

#### Scenario 2
Status
repo.lock exists and is empty
datastore/LOCK exists and is empty
api exists and is non-zero 
daemon is not running

Expected result:
all lock files are removed successfully


## fsrepo/LockedByOtherProcess

I added a debug statement in this function to print out info and found that
calling `lockfile.Locked(repoPath)` twice on the same repo path actually changes
lock status.

The first call returns true, that a lock is held, but the second call (in the
return statement) returns false because the command eventually runs. Why does
this method affect lock status? 

## Test 30 failes

I'm not sure why. 

Thanks!

**Comment by whyrusleeping** [comment]
@pfista Good questions, thanks for the writeup :)

The api file: If the api file doesnt exist, we can assume that no daemon is running.

Good observations on the non-zero repo.lock.. Thats an odd case to handle. There are two different things you could try here.

The first, is the easiest to try. In the place where you set `cannotRunOnDaemon` to true, also set `doesNotUseRepo` to true. That should prevent it from grabbing the repo.lock when it runs the command, although i'm not sure what other side effects this has in your case.

The second thing you could try is to move the entire command from a `Run` action, to a `PreRun` action. The function signature is slightly different for `PreRun`: `func(req cmds.Request) error` and should ensure your code is run before doing a few other potentially troublesome checks. Note: if you do take that route, make sure that the `Run` field doesnt get left empty, assign it a function with nothing in it (so the commands lib code will still actually execute it). Also, in a `PreRun` you can simply use `fmt.Printf` for output to the user instead of setting a response.

**Comment by pfista** [comment]
@whyrusleeping Since we can assume that no daemon is running if there isn't an `api` file, I've rewritten the tests to exclude having an `api` file present when the daemon isn't running. The reason this got tricky is seen in the cases below (where the daemon has not been initialized in the sharness test):

1. In the case that an api file is present and the repo.lock file is empty, my tests would pass.
2. In the case that an api file is present and repo.lock file in not empty, all my tests would fail because `lock.go` errors out when trying to acquire a lock on a non-empty file.

For case number two, instead of changing the command from run to pre-run, or adding the other constraints you mentioned like `doesNotUseRepo`, it seems like the best course of action is to just *not* include an `api` file when testing with a non-zero `repo.lock` file. In the case that I do include the `api` file with a nonzero `repo.lock`, `lock.go` will inevitably check `repo.lock` and always error since it is non-zero. 

So I'm considering this just a poorly written test rather than a bug in the code. That is, *unless* you want to test for this case:

* Daemon is not initialized in sharness test
* api file exists
* repo.lock is non-zero

**Comment by whyrusleeping** [comment]
I think skipping that test is fine. I don't think we need to worry about the case where the daemon isnt initialized and an api file exists.

File: repo/fsrepo/fsrepo.go [hunk 7]
```diff
@@ -259,8 +259,17 @@ func Remove(repoPath string) error {
...
-	// NB: the lock is only held when repos are Open
-	return lockfile.Locked(repoPath)
+	locked, err := lockfile.Locked(repoPath)
+	if locked {
+		log.Debugf("(%t)<->Lock is held at %s", locked, repoPath)
+	}
+	return locked, err
+
+	//if locked, _ := lockfile.Locked(repoPath); locked {
+	//log.Debugf("(%t)<->Lock is held at %s", locked, repoPath)
```
Comment (whyrusleeping) [inline repo/fsrepo/fsrepo.go]: go ahead and remove the commented out code here
